src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -273,14 +223,9 @@ private static string GetTestAgentExePath(bool requires32Bit)
             return Path.Combine(engineDir, agentName);
         }
 
-        private void OnAgentExit(object sender, EventArgs e)
+        private void OnAgentExit(Process process, Guid agentId)
         {
-            var process = sender as Process;
-            if (process == null)
-                return;
-
-            var agentRecord = _agentData.GetDataForProcess(process);
-            agentRecord.Status = AgentStatus.Terminated;
```

Comment:
Hmm...there's a little change of behaviour here isn't there, from a silent failure to a crash, if `OnAgentExit` is every called with a null Process. Are we happy that would never be the case? 🤔 

I'm wandering about things like the issue on mono [here](https://github.com/nunit/nunit-console/issues/676), where it appears the agent is crashing, but the exit code error message is never printed, which suggests the OnAgentExit method is either not being called, or is being called with a null process. I'm not sure we really have testing to cover that sort of area, do we?

Blegh...mono. :-p

Reply:
Sharp. Sorry I forgot to mention it! I did think about this. It is a framework bug for `sender` to not be the same Process instance that you use to subscribe to the event.

> which suggests the OnAgentExit method is either not being called, or is being called with a null process

I'd prefer either a NullReferenceException or some more specific message rather than silently ignoring. That way we find out for sure if there is a buggy framework that doesn't follow the pattern of `sender` matching the instance that raised the event?

Reply:
Agree with the principle - I'd just prefer we tested that on all supported platforms before we put it out in a release...and I'm not sure whether our CI will or not!

I've just pushed a temp branch to simulate an error here, let's see if CI fails. If it does, this sounds good to go - if not, I'd be keen to try and work out a way to test this on mono first.
https://github.com/nunit/nunit-console/compare/tmpCItest?expand=1

Of course, the real solution to this is to have better CI testing. 😄

Reply:
Ok, this was a yak. Turns out exceptions thrown in this method are lost, rendering the entire method redundant...

Separate issue, I think. Also, it's 1.30am here...so tomorrows problem now! Let's get this merged. 🙂

Reply:
There are so many Mono versions though, each with its own idiosyncrasies. Let me know and I'd be happy to put this back.

Reply:
^ commented without seeing your comment. Whatever you are comfortable with!